### ISSUE TITLE
Improve reference_type specification

### ISSUE THREAD

**Issue by mattkretz** [mattkretz/wg21-papers#11]
@jensmaurer:
> (21) I couldn't find a specification for the semantics of reference_type (used for vector elements in operator[]).

There is wording in [datapar.subscr]:
> `reference operator[](size_type i);`
> Returns: A temporary object with the following properties:
> • Remarks: The object is neither DefaultConstructible, CopyConstructible, MoveConstructible, Copy-
Assignable, nor MoveAssignable.
> • Remarks: Assignment, compound assignment, increment, and decrement operators only participate
in overload resolution if called in rvalue context and the corresponding operator of type value_-
type is usable.
> • Effects: The assignment, compound assignment, increment, and decrement operators execute the
indicated operation on the i-th element in the datapar object.
> • Effects: Conversion to value_type returns a copy of the i-th element.

Should I move this to a different place? Should I rather spell out the type? BTW, this spells out the semantics of subscripts on Clang vector builtins.

**Comment by jensmaurer** [comment]
Ah.

The nested "Effects" etc. bullets are ... novel, I believe.

In the class synopsis, I'd say "see below" and then have separate prose text after the class synopsis explaining that "reference" is an implementation-defined type with the following properties. I would not use "CopyConstructible" in the negative, because CopyConstructible comes with a syntactic requirement and with a semantic requirement, i.e. that the copy is the same as the original. Negating "CopyConstructible" is unclear which part is negated. Instead, I'd suggest to say "deleted default constructor, copy constructor, copy assignment operator". This automatically makes the move ctor/assignment go away.
I don't know what "rvalue context" means here, and since fairly recently, we just have a prvalue here and not a temporary object; the temporary object only exists once its materialized (4.4 conv.rval). And which value_type permissible for datapar would make one of the operators mentioned not usable?

**Comment by mattkretz** [comment]
:-D OK.

"rvalue context" was my attempt at saying the operator must have an rvalue-ref qualifier:
```cpp
  reference &operator=(T) &&;
```

E.g. `operator %=` is not usable for integral types. I.e. this is about the compound assignment operators.

**Comment by mattkretz** [comment]
I just tried to split out the `reference` specification (IIUC you'd like me to change the name to `reference_type`, right?). But I don't know how to connect the `i` argument to `operator[]` with the effect a non-const operator call on `reference_type` has. So far it says:

*Effects:* The assignment, compound assignment, increment, and decrement operators execute the indicated operation on the i-th element in the `datapar` object.

> I don't know what "rvalue context" means here, and since fairly recently, we just have a prvalue here and not a temporary object; the temporary object only exists once its materialized (4.4 conv.rval).

Basically I want to make the following ill-formed:
```cpp
datapar<int> v = ...;
auto &s = v[0];
++s;  // error: called on lvalue reference
++v[0];  // OK: called on rvalue reference
```

**Comment by mattkretz** [comment]
@jensmaurer please take a look at ee4614f3e017ae88c94a98caafa9cbb745b82bd7. Is this what you had in mind? Better ideas?

**Comment by jensmaurer** [comment]
The rename to reference_type was wrong (on my part); we seem to call these things simply "reference".

The "rvalue" reformulation looks much better now.

**Comment by mattkretz** [comment]
OK, I consider this resolved then.
